### PR TITLE
feat(Dropdown): Allow freeform to better support time picker scenario

### PR DESCRIPTION
**Pull request by jurokapsiar**
## Current Behavior

- searchable dropdown did not scroll to highlighted item when reopened using toggle button click
- it was not possible to do autocomplete scenarios and support selecting values that are not in the items list
## New Behavior
Intended for time picker scenario:
- when users type a partial string like `11:`, they can see 11:00 selected and then either press `Enter` or `Tab` to select it, or use arrow keys to move around and select
- when users type something that is not in the dropdown, value will still get accepted, for example `11:55`. It is then responsibility of the consumer to validate that string.

Example:
```
import React from "react";
import { Dropdown } from "@fluentui/react-northstar";

const inputItems = [...Array(24).keys()].map((h) => `${h}:00`);

const DropdownExampleSearch = () => {
  const [searchQuery, setSearchQuery] = React.useState("16:00");

  const handleSearchQueryChange = (e, data) => {
    setSearchQuery(data.searchQuery);
  };

  return (
    <>
      <Dropdown
        defaultHighlightedIndex={16}
        onSearchQueryChange={handleSearchQueryChange}
        searchQuery={searchQuery}
        allowFreeform
        search={() => inputItems}
        items={inputItems}
        placeholder="Start typing a name"
        noResultsMessage="We couldn't find any matches."
        getA11ySelectionMessage={{
          onAdd: (item) => `${item} has been selected.`
        }}
      />
      <br />
      Selected value: {searchQuery}
    </>
  );
};

export default DropdownExampleSearch;

```

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/99cc385631d04a76ee10ebc143cb9fecd99640b6'>99cc385631d04a76ee10ebc143cb9fecd99640b6</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## Perf Analysis (`@fluentui/react-northstar`)
:warning: 1 potential perf regressions detected

### Potential regressions comparing to master

Scenario | Current PR Ticks | Baseline Ticks | Ratio | Regression Analysis
:--- | ---:| ---:| ---: | ---: 
LayoutMinimalPerf.default | [264](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LayoutMinimalPerf.default.html) | [259](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LayoutMinimalPerf.default_base.html) | 1.02:1 | [analysis](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LayoutMinimalPerf.default.regression.txt)
<details><summary>Perf tests with no regressions</summary>

Scenario | Current PR Ticks | Baseline Ticks | Ratio
:--- | ---:| ---:| ---:
TableMinimalPerf.default | [341](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TableMinimalPerf.default.html) | [291](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TableMinimalPerf.default_base.html) | 1.17:1
AccordionMinimalPerf.default | [121](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AccordionMinimalPerf.default.html) | [104](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AccordionMinimalPerf.default_base.html) | 1.16:1
LoaderMinimalPerf.default | [571](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LoaderMinimalPerf.default.html) | [503](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LoaderMinimalPerf.default_base.html) | 1.14:1
ListWith60ListItems.default | [517](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListWith60ListItems.default.html) | [460](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListWith60ListItems.default_base.html) | 1.12:1
RadioGroupMinimalPerf.default | [363](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/RadioGroupMinimalPerf.default.html) | [323](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/RadioGroupMinimalPerf.default_base.html) | 1.12:1
VideoMinimalPerf.default | [554](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/VideoMinimalPerf.default.html) | [496](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/VideoMinimalPerf.default_base.html) | 1.12:1
ProviderMinimalPerf.default | [332](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ProviderMinimalPerf.default.html) | [302](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ProviderMinimalPerf.default_base.html) | 1.1:1
HeaderSlotsPerf.default | [603](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/HeaderSlotsPerf.default.html) | [552](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/HeaderSlotsPerf.default_base.html) | 1.09:1
PortalMinimalPerf.default | [140](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/PortalMinimalPerf.default.html) | [128](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/PortalMinimalPerf.default_base.html) | 1.09:1
ItemLayoutMinimalPerf.default | [943](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ItemLayoutMinimalPerf.default.html) | [877](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ItemLayoutMinimalPerf.default_base.html) | 1.08:1
ButtonMinimalPerf.default | [122](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ButtonMinimalPerf.default.html) | [114](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ButtonMinimalPerf.default_base.html) | 1.07:1
GridMinimalPerf.default | [258](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/GridMinimalPerf.default.html) | [241](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/GridMinimalPerf.default_base.html) | 1.07:1
AttachmentMinimalPerf.default | [116](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AttachmentMinimalPerf.default.html) | [109](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AttachmentMinimalPerf.default_base.html) | 1.06:1
MenuButtonMinimalPerf.default | [1395](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/MenuButtonMinimalPerf.default.html) | [1321](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/MenuButtonMinimalPerf.default_base.html) | 1.06:1
AvatarMinimalPerf.default | [139](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AvatarMinimalPerf.default.html) | [133](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AvatarMinimalPerf.default_base.html) | 1.05:1
FormMinimalPerf.default | [304](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/FormMinimalPerf.default.html) | [289](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/FormMinimalPerf.default_base.html) | 1.05:1
LabelMinimalPerf.default | [312](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LabelMinimalPerf.default.html) | [297](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/LabelMinimalPerf.default_base.html) | 1.05:1
ListNestedPerf.default | [430](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListNestedPerf.default.html) | [410](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListNestedPerf.default_base.html) | 1.05:1
SkeletonMinimalPerf.default | [264](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SkeletonMinimalPerf.default.html) | [252](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SkeletonMinimalPerf.default_base.html) | 1.05:1
ListMinimalPerf.default | [399](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListMinimalPerf.default.html) | [383](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListMinimalPerf.default_base.html) | 1.04:1
StatusMinimalPerf.default | [505](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/StatusMinimalPerf.default.html) | [486](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/StatusMinimalPerf.default_base.html) | 1.04:1
TooltipMinimalPerf.default | [929](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TooltipMinimalPerf.default.html) | [897](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TooltipMinimalPerf.default_base.html) | 1.04:1
CardMinimalPerf.default | [449](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CardMinimalPerf.default.html) | [438](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CardMinimalPerf.default_base.html) | 1.03:1
CarouselMinimalPerf.default | [354](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CarouselMinimalPerf.default.html) | [343](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CarouselMinimalPerf.default_base.html) | 1.03:1
ImageMinimalPerf.default | [278](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ImageMinimalPerf.default.html) | [269](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ImageMinimalPerf.default_base.html) | 1.03:1
RefMinimalPerf.default | [199](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/RefMinimalPerf.default.html) | [194](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/RefMinimalPerf.default_base.html) | 1.03:1
BoxMinimalPerf.default | [242](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/BoxMinimalPerf.default.html) | [238](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/BoxMinimalPerf.default_base.html) | 1.02:1
ChatMinimalPerf.default | [546](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ChatMinimalPerf.default.html) | [534](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ChatMinimalPerf.default_base.html) | 1.02:1
DatepickerMinimalPerf.default | [4549](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DatepickerMinimalPerf.default.html) | [4450](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DatepickerMinimalPerf.default_base.html) | 1.02:1
ChatWithPopoverPerf.default | [284](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ChatWithPopoverPerf.default.html) | [282](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ChatWithPopoverPerf.default_base.html) | 1.01:1
DropdownManyItemsPerf.default | [499](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DropdownManyItemsPerf.default.html) | [492](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DropdownManyItemsPerf.default_base.html) | 1.01:1
DropdownMinimalPerf.default | [2403](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DropdownMinimalPerf.default.html) | [2387](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DropdownMinimalPerf.default_base.html) | 1.01:1
EmbedMinimalPerf.default | [3173](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/EmbedMinimalPerf.default.html) | [3134](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/EmbedMinimalPerf.default_base.html) | 1.01:1
HeaderMinimalPerf.default | [256](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/HeaderMinimalPerf.default.html) | [254](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/HeaderMinimalPerf.default_base.html) | 1.01:1
InputMinimalPerf.default | [995](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/InputMinimalPerf.default.html) | [981](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/InputMinimalPerf.default_base.html) | 1.01:1
MenuMinimalPerf.default | [641](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/MenuMinimalPerf.default.html) | [635](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/MenuMinimalPerf.default_base.html) | 1.01:1
ProviderMergeThemesPerf.default | [947](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ProviderMergeThemesPerf.default.html) | [942](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ProviderMergeThemesPerf.default_base.html) | 1.01:1
ReactionMinimalPerf.default | [271](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ReactionMinimalPerf.default.html) | [267](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ReactionMinimalPerf.default_base.html) | 1.01:1
SplitButtonMinimalPerf.default | [3424](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SplitButtonMinimalPerf.default.html) | [3382](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SplitButtonMinimalPerf.default_base.html) | 1.01:1
AttachmentSlotsPerf.default | [804](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AttachmentSlotsPerf.default.html) | [805](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AttachmentSlotsPerf.default_base.html) | 1:1
ButtonOverridesMissPerf.default | [1118](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ButtonOverridesMissPerf.default.html) | [1113](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ButtonOverridesMissPerf.default_base.html) | 1:1
FlexMinimalPerf.default | [204](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/FlexMinimalPerf.default.html) | [205](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/FlexMinimalPerf.default_base.html) | 1:1
CustomToolbarPrototype.default | [2085](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CustomToolbarPrototype.default.html) | [2095](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CustomToolbarPrototype.default_base.html) | 1:1
RosterPerf.default | [859](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/RosterPerf.default.html) | [872](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/RosterPerf.default_base.html) | 0.99:1
SegmentMinimalPerf.default | [246](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SegmentMinimalPerf.default.html) | [248](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SegmentMinimalPerf.default_base.html) | 0.99:1
SliderMinimalPerf.default | [1271](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SliderMinimalPerf.default.html) | [1286](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/SliderMinimalPerf.default_base.html) | 0.99:1
TextAreaMinimalPerf.default | [352](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TextAreaMinimalPerf.default.html) | [354](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TextAreaMinimalPerf.default_base.html) | 0.99:1
ToolbarMinimalPerf.default | [714](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ToolbarMinimalPerf.default.html) | [718](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ToolbarMinimalPerf.default_base.html) | 0.99:1
AnimationMinimalPerf.default | [450](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AnimationMinimalPerf.default.html) | [457](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AnimationMinimalPerf.default_base.html) | 0.98:1
ChatDuplicateMessagesPerf.default | [211](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ChatDuplicateMessagesPerf.default.html) | [215](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ChatDuplicateMessagesPerf.default_base.html) | 0.98:1
PopupMinimalPerf.default | [472](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/PopupMinimalPerf.default.html) | [483](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/PopupMinimalPerf.default_base.html) | 0.98:1
TreeMinimalPerf.default | [648](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TreeMinimalPerf.default.html) | [663](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TreeMinimalPerf.default_base.html) | 0.98:1
ButtonSlotsPerf.default | [404](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ButtonSlotsPerf.default.html) | [415](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ButtonSlotsPerf.default_base.html) | 0.97:1
ListCommonPerf.default | [505](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListCommonPerf.default.html) | [521](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/ListCommonPerf.default_base.html) | 0.97:1
IconMinimalPerf.default | [422](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/IconMinimalPerf.default.html) | [436](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/IconMinimalPerf.default_base.html) | 0.97:1
TableManyItemsPerf.default | [1453](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TableManyItemsPerf.default.html) | [1498](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TableManyItemsPerf.default_base.html) | 0.97:1
CheckboxMinimalPerf.default | [2034](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CheckboxMinimalPerf.default.html) | [2108](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/CheckboxMinimalPerf.default_base.html) | 0.96:1
DialogMinimalPerf.default | [564](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DialogMinimalPerf.default.html) | [597](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DialogMinimalPerf.default_base.html) | 0.94:1
TreeWith60ListItems.default | [125](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TreeWith60ListItems.default.html) | [134](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TreeWith60ListItems.default_base.html) | 0.93:1
TextMinimalPerf.default | [256](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TextMinimalPerf.default.html) | [286](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/TextMinimalPerf.default_base.html) | 0.9:1
DividerMinimalPerf.default | [250](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DividerMinimalPerf.default.html) | [290](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/DividerMinimalPerf.default_base.html) | 0.86:1
AlertMinimalPerf.default | [186](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AlertMinimalPerf.default.html) | [228](https://fluentuipr.z22.web.core.windows.net/pull/23949/perf-test-northstar/AlertMinimalPerf.default_base.html) | 0.82:1

</details><div id="perfComment9424"/>

**Comment by ling1726** [comment]
In this case the a11y selection status message will always be wrong, regardless if the user applies validation right ?

If I type 'foo', the narrated message will be '0:00 has been selected'

File: packages/fluentui/react-northstar/src/components/Dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -911,8 +928,19 @@ export const Dropdown = (React.forwardRef<HTMLDivElement, DropdownProps>((props,
         const newValue = isSameItemSelected ? value[0] : changes.selectedItem;
 
         newState.searchQuery = getSelectedItemAsString(newValue);
+        if (allowFreeform && !inListbox.current && type === Downshift.stateChangeTypes.keyDownEnter) {
```
[the remark this answers is not in the document]
Comment (jurokapsiar) [reply packages/fluentui/react-northstar/src/components/Dropdown/Dropdown.tsx:931]: it is, but in different part of the code as tab is not based on downshift state change

File: packages/fluentui/react-northstar/src/components/Dropdown/Dropdown.tsx [hunk 2]
```diff
@@ -911,8 +928,19 @@ export const Dropdown = (React.forwardRef<HTMLDivElement, DropdownProps>((props,
...
+          const itemIndex = items.findIndex(i => itemToString(i)?.startsWith(searchQuery));
+
+          // if there is an item that starts with searchQuery, still apply the search query
+          // to do auto complete (you enter '12:', can be completed to '12:00')
+          if (itemIndex === -1) {
+            delete newState.searchQuery;
```
[the remark this answers is not in the document]
Comment (jurokapsiar) [reply packages/fluentui/react-northstar/src/components/Dropdown/Dropdown.tsx:937]: no, in this case search query should not change, we should keep the previous value. the comment above explains it, do we need to improve it?

**Review by ling1726 — APPROVED** [review]

**Pull request merged**